Update: fix typo `from_pretained` to `from_pretrained`

YooSungHyun opened the pull request:
# What does this PR do?
Fix typo `from_pretained` to `from_pretrained`

Fixes # (issue)
https://github.com/huggingface/optimum/issues/1088

## Before submitting
- [x] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests?

chk plz! @michaelbenayoun @jin

YooSungHyun commented:
"maybe this is lookin small, but can made big wave...! 😂"

michaelbenayoun commented:
@YooSungHyun Let's talk about the issues here instead than #1088.

About quantization attribute not being a `QuantizationConfig` but a dictionary, you are right, and this is bad because it is actually documented as a `QuantizationConfig`.

You can also fix this here, by doing in `optimum/onnxruntime/configuration.py`:

```python
def __init__(                                                                                                                                                                           
    self,                                                                                                                                                                                         
    opset: Optional[int] = None,                                                                                                                                                                  
    use_external_data_format: bool = False,                                                                                                                                             
    one_external_file: bool = True,                                                                                                                                                     
    optimization: Optional[OptimizationConfig] = None,                                                                                                                                  
    quantization: Optional[QuantizationConfig] = None,                                                                                                                                  
    **kwargs,                                                                                                                                                                           
      ): -> None                                                                                                                                                                              
          super().__init__()                                                                                                                                                                  
          self.opset = opset                                                                                                                                                   
          self.use_external_data_format = use_external_data_format                                                                                                                   
          self.one_external_file = one_external_file                                                                                                                      
          self.optimization = optimization                                                                                                                      
          self.quantization = quantization                                                                                                                     
          self.optimum_version = kwargs.pop("optimum_version", None)    
```

Basically the only difference here is that we do not call the `dataclass_to_dict` method.

Then to be able to serialize `ORTConfig`, you will need to override the `to_dict` method:

```python
def to_dict(self) -> Dict[str, Any]:
    clone = copy.deepcopy(self) : Self@ORTConfig                                                                                                                                        
    clone.optimization = self.dataclass_to_dict(clone.optimization)
    clone.quantization = self.dataclass_to_dict(clone.quantization) : dict[Unknown, Unknown]                                                                                            
    return BaseConfig.to_dict(clone)
```

YooSungHyun commented:
@michaelbenayoun you mean, convert to `OptimizationConfig` and `QuantizationConfig` in `__init__():` and if when i needs some `save_pretrained()` kind of thing, then `ORTConfig().to_dict()` -> `save_pretrained()`??

michaelbenayoun commented:
No I mean:

1. No transforming the `optimization` and `quantization` attributes to dictionaries in `ORTConfig.__init__`.
2. Doing that will make the `ORTConfig` not JSON-serializable, to make it serializable, just add the `to_dict` method I shared to the `ORTConfig` class.

YooSungHyun commented:
first of all, i commit first. so let we talk with `code conversation`..!

YooSungHyun commented:
@michaelbenayoun 
first, i don't speak english as well. so i may not understand it very well..but, I want to contribute this code intensely..! so.... plz bear with me🤗

YooSungHyun commented:
i will convert source code about `ORTConfig.from_pretrained()` ->  `ORTConfig.from_dict()`
i will override `PretrainedConfig.from_dict()` in `ORTConfig`. so, It will just call `ORTConfig.from_pretrained()` then, `optimization` and `quantization` attribute will be each `OptimizationConfig` and `Quantizationconfig`

michaelbenayoun commented on optimum/onnxruntime/configuration.py:
```diff
@@ -612,6 +613,40 @@ def tensorrt(
             qdq_dedicated_pair=True,
         )
 
+    @staticmethod
+    def from_config(config: dict) -> QuantizationConfig:
+        """
+        Creates a [`~onnxruntime.QuantizationConfig`] from input config json file (default: ort_config.json)
+        Args:
+            config (`dict`):
+                Input ONNX Runtime config file's (e.g. ort_config.json) `quantization` attribute
+        """
+        format, mode = default_quantization_parameters(is_static=config["is_static"])
+        if config["activations_dtype"].lower() in ["u8", "quint8", "uint8"]:
+            activations_dtype = QuantType.QUInt8
+        else:
+            activations_dtype = QuantType.QInt8
+        if config["weights_dtype"].lower() in ["u8", "quint8", "uint8"]:
+            weights_dtype = QuantType.QUInt8
+        else:
+            weights_dtype = QuantType.QInt8
+        return QuantizationConfig(
+            is_static=config["is_static"],
+            format=format,
+            mode=mode,
+            activations_dtype=activations_dtype,
+            activations_symmetric=config["activations_symmetric"],
+            weights_dtype=weights_dtype,
+            weights_symmetric=config["weights_symmetric"],
+            per_channel=config["per_channel"],
+            reduce_range=config["reduce_range"],
+            nodes_to_quantize=config["nodes_to_quantize"] or [],
+            nodes_to_exclude=config["nodes_to_exclude"] or [],
+            operators_to_quantize=config["operators_to_quantize"] or ["MatMul", "Add"],
+            qdq_add_pair_to_weight=config["qdq_add_pair_to_weight"],
+            qdq_dedicated_pair=config["qdq_dedicated_pair"],
+        )
+
```
Why?

YooSungHyun commented:
@michaelbenayoun  Ta-da! i've changed all my code.
i think, before version source-code is two-step necessary.
1. load all ort config
2. make QuantizationConfig
and this way, if someone needs OptimizationConfig, he or she must write down OptimizationConfig's load code. but, QuantizationConfig is not have any side-effect

but, look at after version...
just override `from_dict` for `BaseConfig.from_pretrained()`, it can will be load all each config classes (OptimizationConfig and QuantizationConfig) at once.
and also, using `getattr`, we can load exactly type (if is none, will be default value)
if, someone needs OptimizationConfig, he or she just write down under the `if config.optimization:`, or maybe all loaded well.

finally, at `quantize.py` it is more scalable, because just using ORTConfig is ORTConfig, not any additional change (what if using additional config value)

i deleted `super.__init__()` that is make not used additional parameter
and i think `pruned_heads` and some like `label` kind of things is not using ORTConfig, so deleted too.

this way, is using `override`, so i think have some side-effect problem. so code review is necessary

i tested it well done, and when if config json key missing, it will set default value...
how about it?

michaelbenayoun commented on optimum/onnxruntime/configuration.py:
```diff
@@ -965,3 +965,66 @@ def dataclass_to_dict(config) -> dict:
+        # Those arguments may be passed along for our internal telemetry.
+        # We remove them so they don't appear in `return_unused_kwargs`.
+        kwargs.pop("_from_auto", None)
+        kwargs.pop("_from_pipeline", None)
+        # The commit hash might have been updated in the `config_dict`, we don't want the kwargs to erase that update.
+        if "_commit_hash" in kwargs and "_commit_hash" in config_dict:
+            kwargs["_commit_hash"] = config_dict["_commit_hash"]
+
+        config: ORTConfig = cls(**config_dict)
+
+        if config.optimization:
+            config.optimization: OptimizationConfig = OptimizationConfig(**config.optimization)
+
+        if config.quantization:
+            quant_config: dict = config.quantization
+            quant_config["format"], quant_config["mode"] = default_quantization_parameters(
+                is_static=quant_config["is_static"],
+                format=getattr(QuantFormat, quant_config.get("format", ""), None),
+                mode=getattr(QuantizationMode, quant_config.get("mode", ""), None),
+            )
+            quant_config["activations_dtype"] = getattr(
+                QuantType, quant_config.get("activations_dtype", ""), QuantType.QUInt8
+            )
+            quant_config["weights_dtype"] = getattr(QuantType, quant_config.get("weights_dtype", ""), QuantType.QInt8)
+            config.quantization: QuantizationConfig = QuantizationConfig(**quant_config)
+
+        to_remove = []
+        for key, value in kwargs.items():
+            if hasattr(config, key):
+                setattr(config, key, value)
+                if key != "torch_dtype":
+                    to_remove.append(key)
+        for key in to_remove:
+            kwargs.pop(key, None)
+
+        logger.info(f"ORTConfig: {config.to_dict()}")
+        if return_unused_kwargs:
+            return config, kwargs
+        else:
+            return config
```
I think a better approach would be to do:

```suggestion
    def from_dict(cls, config_dict: Dict[str, Any], **kwargs) -> "PretrainedConfig":
         config = super().from_dict(config_dict, **kwargs)
         if config.quantization:
             config.quantization = QuantizationConfig(**config.quantization)
         else:
             config.quantization = None
         # Same thing for optimization  
```

This way we duplicate less code.

YooSungHyun commented:
[69ba1c7](https://github.com/huggingface/optimum/pull/1089/commits/69ba1c76dec1676c997559d4e0bb61fad5bfd30c)

is good i think. how about that?

@michaelbenayoun

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/optimum/pr_1089). All of your documentation changes will be reflected on that endpoint.

YooSungHyun commented:
@fxmarty @michaelbenayoun 🤗

fxmarty commented:
Thank you @YooSungHyun and apologizes for the delay, I'll give a review on Monday!

fxmarty commented:
Can you run `make style`?

fxmarty commented on optimum/commands/onnxruntime/quantize.py:
```diff
@@ -83,6 +83,7 @@ def run(self):
             for model in self.args.onnx_model.glob("*.onnx")
         ]
 
+        use_external_data_format = False
```
Why is it hard-coded here?

fxmarty commented on optimum/onnxruntime/configuration.py:
```diff
@@ -343,6 +344,13 @@ def ensure_valid_data_type_or_raise(
         )
 
+def set_quantization_parameters(is_static, format, mode, activations_dtype, weights_dtype):
+    format = getattr(QuantFormat, str(format or ""), None),
```
Why `str(format or "")` (here and the following)?

fxmarty commented on optimum/onnxruntime/configuration.py:
```diff
@@ -343,6 +344,13 @@ def ensure_valid_data_type_or_raise(
             "activations_dtype = QuantType.QInt8 with weights_dtype = QuantType.QUInt8."
         )
 
+def set_quantization_parameters(is_static, format, mode, activations_dtype, weights_dtype):
```
It could be nice to have type hints here.

fxmarty commented on optimum/onnxruntime/configuration.py:
```diff
@@ -943,12 +951,11 @@ def __init__(
         quantization: Optional[QuantizationConfig] = None,
```
So are these type hints correct in the end? `optimization: Optional[OptimizationConfig] = None,` and `quantization: Optional[QuantizationConfig] = None,`? According to https://github.com/huggingface/optimum/pull/1089#issuecomment-1582233708 they are not?

fxmarty replied:
bump

michaelbenayoun commented on optimum/onnxruntime/configuration.py:
```diff
@@ -344,6 +345,41 @@ def ensure_valid_data_type_or_raise(
+    is_static: bool, format: str = "", mode: str = "", activations_dtype: str = "", weights_dtype: str = ""
+):
+    """
+    Get each QuantFormat, QuantizationMode, QuantType to str in ort_config.json
```
```suggestion
    Converts `QuantFormat`, `QuantizationMode`, `QuantType` to string for serialization.
```

michaelbenayoun commented on optimum/onnxruntime/configuration.py:
```diff
@@ -344,6 +345,41 @@ def ensure_valid_data_type_or_raise(
+):
+    """
+    Get each QuantFormat, QuantizationMode, QuantType to str in ort_config.json
+    So, each onnx runtime config parameters is had to name each classes's class attribute names
```
I do not understand.

michaelbenayoun commented on optimum/onnxruntime/configuration.py:
```diff
@@ -344,6 +345,41 @@ def ensure_valid_data_type_or_raise(
+    Get each QuantFormat, QuantizationMode, QuantType to str in ort_config.json
+    So, each onnx runtime config parameters is had to name each classes's class attribute names
+
+    Example) onnxruntime/onnxruntime/python/tools/quantization/quant_utils.py
```
```suggestion
    Example: onnxruntime/onnxruntime/python/tools/quantization/quant_utils.py
```

michaelbenayoun commented on optimum/onnxruntime/configuration.py:
```diff
@@ -344,6 +345,41 @@ def ensure_valid_data_type_or_raise(
+        - QuantizationMode.IntegerOps   = "IntegerOps"
+        - QuantizationMode.QLinearOps   = "QLinearOps"
+        - QuantType.QInt8               = "QInt8"
+        - QuantType.QUInt8              = "QUInt8"
+    Args:
```
```suggestion
        - QuantType.QUInt8              = "QUInt8"
        
    Args:
```

michaelbenayoun commented on optimum/onnxruntime/configuration.py:
```diff
@@ -965,3 +1000,34 @@ def dataclass_to_dict(config) -> dict:
+            config.optimization: OptimizationConfig = OptimizationConfig(**config.optimization)
+
+        if config.quantization:
+            quant_config: dict = config.quantization
+            (
+                quant_config["format"],
+                quant_config["mode"],
+                quant_config["activations_dtype"],
+                quant_config["weights_dtype"],
+            ) = set_quantization_parameters(
+                quant_config["is_static"],
+                quant_config.get("format", ""),
+                quant_config.get("mode", ""),
+                quant_config.get("activations_dtype", ""),
+                quant_config.get("weights_dtype", ""),
+            )
```
Not very readable IMO

YooSungHyun commented:
@fxmarty

YooSungHyun commented:
@fxmarty @michaelbenayoun  plz check...!😅
